### PR TITLE
Add `find_successors_by_edge`

### PR DESCRIPTION
**Pull request by IvanIsCoding**
Related to #265

This PR adds two new functions to `PyDiGraph`: `find_sucessors_by_edge` and `find_predecessors_by_edge`. These functions can be useful for speeding up routines that want a subset of successors/predecessors that match a filter function.

With these new functions, it is possible to find the successors/predecessors of an node that match a filter in `O(e')`, where `e'` is the number of edges connected to the node. Before, the simplest way to achieve the same behavior was to get all edges for each successor/predecessor and then filter, which was `O(d * e')`, where `d` is the number of distinct neighbors a node has.

These new functions will be immediatelly applicable to the `quantum_successors` and `quantum_precessors` methods in `qiskit-terra` [DAGCircuit](https://github.com/Qiskit/qiskit-terra/blob/9b7893439c8e1c58edf0b3829e09c984add0a122/qiskit/dagcircuit/dagcircuit.py#L1272). They will consequently provide a speed up to [collect 2q blocks](https://github.com/Qiskit/qiskit-terra/blob/b058856e2aee9c5819120dc9c95f1e2478c95f4e/qiskit/transpiler/passes/optimization/collect_2q_blocks.py#L55-L246), which has `quantum_successors` and `quantum_precessors` as its most expensive steps during graph traversal.

### CODE REVIEW

**Review by nahumsa — APPROVED** [review]
LGTM. Very clean implementation and tests. 
Just one question: have you benchmarked this improvement comparing with the previous implementation?

**Comment by IvanIsCoding** [comment]
> LGTM. Very clean implementation and tests.
> Just one question: have you benchmarked this improvement comparing with the previous implementation?

I have benchmarks for the code that uses it in `qiskit-terra`, there was an overall improvement there. I will edit this with a specific benchmark for this new function.

File: releasenotes/notes/find-successors-by-d3589b1b1d0b1633.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+features:
+  - |
+    Two new functions were added to :class:`~retworkx.PyDiGraph`: :func:`~retworkx.PyDiGraph.find_successors_by_edge`
```
Comment (mtreinish) [inline releasenotes/notes/find-successors-by-d3589b1b1d0b1633.yaml:4]:
```suggestion
    Two new functions were added to :class:`~retworkx.PyDiGraph`: :meth:`~retworkx.PyDiGraph.find_successors_by_edge`
```

File: releasenotes/notes/find-successors-by-d3589b1b1d0b1633.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+    and :func:`~retworkx.PyDiGraph.find_predecessors_by_edge` to efficiently retrieve the neighbors in the graph
```
Comment (mtreinish) [inline releasenotes/notes/find-successors-by-d3589b1b1d0b1633.yaml:5]:
```suggestion
    and :meth:`~retworkx.PyDiGraph.find_predecessors_by_edge` to efficiently retrieve the neighbors in the graph
```

**Review by mtreinish — APPROVED** [review]
LGTM, thanks for doing this

**Pull request merged**